### PR TITLE
Initial package setup

### PR DESCRIPTION
**Pull request by tomkuschka**
## Description

This PR adds initial package functionality like:
* Adds readme
* Initial project setup
* Unit tests
* Basic token generation functionality
  * ColorScheme
  * TextStyle
  * Theme Extensions

## Checklist

- [x] I added a PR description.
- [x] I linked all related issues and PRs I could find (no links if there are none).
- [x] If this PR changes anything about the main `design_tokens_builder` package
  (also README etc.), I created an entry in `CHANGELOG.md` (`## UPCOMING RELEASE` if the change
  on its own is not worth an update).
- [x] If this PR includes a notable change in the `design_token_builder` package, I updated the version
  according to [Dart's semantic versioning](https://stackoverflow.com/questions/66201337/how-do-dart-package-versions-work-how-should-i-version-my-flutter-plugins/66201338#66201338).
- [x] If there is new functionality in code, I added tests covering all my additions.
- [x] All required checks pass.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+A package for generating a flutter `ThemeData` from an exported design token json that is used e.g. [in this Figma plugin](https://tokens.studio/). It rather aims to create a flutter `Theme` 
+based on Material 3 than creating a lot of custom widgets for each token.
+
+## Features
+
+Right now we support generating/parsing the following:
+- `ColorScheme`
+- `TextStyle`
+  - Font family mapping (from Figma font name to flutter name)
```
Comment (ukasz123) [inline README.md:18]:
```suggestion
  - Font family mapping (from Figma font name to Flutter name)
```

Comment (ukasz123) [reply README.md:18]: There are other occurrences of this typo.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
+<h1 align="center">
+    <picture>
+      <source media="(prefers-color-scheme: dark)" srcset="https://github.com/simpleclub/design_tokens_builder/assets/35028202/0e70832e-78f2-4906-a5b1-940804f2619a">
+      <img alt="Logo" src="https://github.com/simpleclub/design_tokens_builder/assets/35028202/6ec91bfa-3124-4430-9ed0-6d616fe5c838">
+    </picture>
+</h1>
```
Comment (Sese-Schneider) [inline README.md:6]: This looks bad when rendered on github

File: pubspec.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
+name: design_tokens_builder
+description: A new Flutter package project.
```
Comment (Sese-Schneider) [inline pubspec.yaml:2]: Invalid description

File: pubspec.yaml [hunk 4]
```diff
@@ -0,0 +1,22 @@
+name: design_tokens_builder
+description: A new Flutter package project.
+version: 0.0.1
+
+environment:
+  sdk: '>=2.19.5 <3.0.0'
```
Comment (Sese-Schneider) [inline pubspec.yaml:6]: Lets make it also dart 3 compatible

**Review by Sese-Schneider — CHANGES_REQUESTED** [review]
Nice work @tomkuschka 

I would love to see a [CODEOWNERS](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) with me and you as GLOBAL owners added

File: README.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+`ThemeData` should always start with `sys`. If you then e.g. want to set the primary color in flutters `ColorScheme` you
+have to add a token called `sys.primary` with type `color` to your json file containing all of your tokens. `primary` is
+the name of `ColorScheme`s `primary` field. Keep in mind to exactly write the field names in camel case so the package is
+able to recognize them properly. 
+For text styles it works similar. Here also use `sys` as the leading part of your token. To better organize the tokens in
+Token Studio we decided to split the text style naming. So if you want to generate `displaySmall` text style just use a
+`sys.display.small` token of type `typography`.
+
+### Generation of theme extensions
+Theme extension are a great way providing custom tokens that are not in flutters `ThemeData` defined. In order to create
```
Comment (millerovv) [inline README.md:79]:
```suggestion
Theme extensions are a great way of providing custom parameters that are not defined in `ThemeData`. In order to create

```

File: README.md [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+a new extension you can just use a different prefix then `sys` for your tokens. Based on that name and the type of the
+tokens inside that group an extension will get generated and added to the `ThemeData`. So a custom set of colors called
```
Comment (millerovv) [inline README.md:81]: > inside that group

Could you specify which group?

File: README.md [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+### Multi-theming and dark and light mode
+The package can generate multiple themes based on multiple [token sets](https://docs.tokens.studio/themes/token-sets) in
+your token data json. Every set will be available through a `GeneratedTokenSet`. This enum contains all set pairs and 
+`ThemeData` for light and dark themes. In order to specify the brightness of your set use Light/Dark as a suffix in 
+design tokens (e.g. yourSetLight, yourSetDark). You also might have a global or core set with universal tokens that do 
+not change based on the selected brightness. In that case make sure to call this set `global`. It will be recognised by 
+the builder and ignored when creating `GeneratedTokenSets`.
+
+### Generation of flutter `ThemeData`
+The package allows for easy and simple generation of flutters `ThemeData`. For doing so, you have to have a certain
```
Comment (millerovv) [inline README.md:68]:
```suggestion
The package allows for easy and simple generation of Flutter's `ThemeData`. For doing so, you have to have a certain 

```

File: README.md [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+your token data json. Every set will be available through a `GeneratedTokenSet`. This enum contains all set pairs and 
+`ThemeData` for light and dark themes. In order to specify the brightness of your set use Light/Dark as a suffix in 
+design tokens (e.g. yourSetLight, yourSetDark). You also might have a global or core set with universal tokens that do 
+not change based on the selected brightness. In that case make sure to call this set `global`. It will be recognised by 
+the builder and ignored when creating `GeneratedTokenSets`.
+
+### Generation of flutter `ThemeData`
+The package allows for easy and simple generation of flutters `ThemeData`. For doing so, you have to have a certain 
+structure for your design tokens. Since this is a system token related to material 3 the tokens used to generate flutters
+`ThemeData` should always start with `sys`. If you then e.g. want to set the primary color in flutters `ColorScheme` you
```
Comment (millerovv) [inline README.md:70]: Any change the structure and the sys rule can be demonstrated with a code snippet or a screenshot?

**Comment by ukasz123** [comment]
I'm upset about the fact there is no data class model for JSON and everything is dynamically typed so the token generation is so "JS-style" I have no clue what I'm looking at actually.

**Review by Sese-Schneider — APPROVED** [review]

File: pubspec.yaml [hunk 9]
```diff
@@ -0,0 +1,27 @@
+name: design_tokens_builder
+description: >-2
+  Package for generating Flutter ThemeData from Figma design tokens using BuildRunner.
+version: 0.1.0
```
Comment (IchordeDionysos) [inline pubspec.yaml:4]:
```suggestion
version: 0.1.0
homepage: https://github.com/simpleclub/design_tokens_builder
```

File: pubspec.yaml [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+  yaml: ^3.1.2
+
+dev_dependencies:
+  flutter_lints: ^2.0.0
+  flutter_test:
+    sdk: flutter
+
+screenshots:
+  - description: 'Image showing the conversion from a Design Token Studio json to Flutter code.'
+    path: screenshots/conversion.png
```
Comment (IchordeDionysos) [inline pubspec.yaml:27]:
```suggestion
screenshots:
  - description: 'Image showing the conversion from a Design Token Studio json to Flutter code.'
    path: screenshots/conversion.png
topics:
  - Figma
  - Design Tokens
  - Material
  - Theme
  - Generator

```

**Review by Sese-Schneider — APPROVED** [review]

**Review by millerovv — APPROVED** [review]

**Comment by Sese-Schneider** [comment]
@tomkuschka code needs to be formatted

**Review by millerovv — APPROVED** [review]
LGTM!

File: test/utils/token_set_utils_test.dart [hunk 11]
```diff
@@ -0,0 +1,117 @@
+import 'package:design_tokens_builder/utils/token_set_utils.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+void main() {
+  group('Get token sets', () {
+    test('succeeds', () {
+      final result = getTokenSets({
+        '\$metadata': {
```
Comment (IchordeDionysos) [inline test/utils/token_set_utils_test.dart:8]: nit: Easier to read

```suggestion
        r'$metadata': {
```

File: test/utils/token_set_utils_test.dart [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+          'tokenSetOrder': ['global', 'light', 'dark'],
+        },
+      });
+
+      expect(result, ['light', 'dark']);
+    });
+
+    test('succeeds without default theme', () {
+      final result = getTokenSets({
+        '\$metadata': {
```
Comment (IchordeDionysos) [inline test/utils/token_set_utils_test.dart:18]: nit:
```suggestion
        r'$metadata': {
```

File: test/utils/token_set_utils_test.dart [hunk 13]
```diff
@@ -0,0 +1,117 @@
+import 'package:design_tokens_builder/utils/token_set_utils.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+void main() {
+  group('Get token sets', () {
+    test('succeeds', () {
+      final result = getTokenSets({
+        '\$metadata': {
+          'tokenSetOrder': ['global', 'light', 'dark'],
```
Comment (IchordeDionysos) [inline test/utils/token_set_utils_test.dart:9]: So global is always a default token set? Is that something normal in the design token generator?

File: lib/utils/token_set_utils.dart [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+Map<String, dynamic> overrideAndMergeTokenSet(
+  Map<String, dynamic> baseSet, {
+  required Map<String, dynamic> withSet,
+}) {
+  return baseSet..addAll(withSet);
+}
+
+/// Check if `data` has any child that has type `type`.
+bool hasNestedType(MapEntry<String, dynamic> data, {required String type}) {
+  final value = data.value as Map<String, dynamic>;
```
Comment (IchordeDionysos) [inline lib/utils/token_set_utils.dart:22]: What if the value is not a map?

Comment (IchordeDionysos) [reply lib/utils/token_set_utils.dart:22]: Wouldn't it then fail when calling `values.entries` later?

File: lib/utils/token_set_utils.dart [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+  if (value.containsKey('type')) {
+    return value['type'] == type;
+  }
+
+  return hasNestedType(value.entries.first, type: type);
```
Comment (IchordeDionysos) [inline lib/utils/token_set_utils.dart:27]: Is it not required to loop over all entries here?

File: test/utils/token_set_utils_test.dart [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+        'display': {
+          'large': {'value': 'Some value', 'type': 'someType'}
+        },
+      }, fallbackSetData: {
+        'small': {'value': 'Some value', 'type': 'someType'},
+        'large': {'value': 'Some value', 'type': 'someType'},
+      });
+
+      expect(result.keys.length, 3);
+      expect(result['small'], {'value': 'Some new value', 'type': 'someType'});
```
Comment (IchordeDionysos) [inline test/utils/token_set_utils_test.dart:110]: nit: Maybe we should also check medium?
```suggestion
      expect(result['small'], {'value': 'Some new value', 'type': 'someType'});
      expect(result['medium'], null);
```

File: test/utils/typography_utils_test.dart [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+            ],
+          },
+        ),
+      );
+
+      expect(
+          result,
+          'TextStyle(\n'
+          '      fontFamily: \'MyFont\',\n'
+          '      fontWeight: null,\n'
```
Comment (IchordeDionysos) [inline test/utils/typography_utils_test.dart:76]: Why is font weight null here? 🤔

File: test/utils/typography_utils_test.dart [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+          '      height: null,\n'
```
Comment (IchordeDionysos) [inline test/utils/typography_utils_test.dart:77]: Does height refer to line height here and if yes why is it null?

**Review by IchordeDionysos — COMMENTED** [review]
Just a bit nitpicking nothing blocking 😌

File: lib/utils/token_set_utils.dart [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+}
```
[the remark this answers is not in the document]
Comment (tomkuschka) [reply lib/utils/token_set_utils.dart:28]: Will be fixed with next PR

File: lib/utils/token_set_utils.dart (hunk 14 above)
Comment (tomkuschka) [reply lib/utils/token_set_utils.dart:22]: Yes true. Will be fixed with next PR. (There its already fixed)

File: test/utils/typography_utils_test.dart (hunk 17 above)
Comment (tomkuschka) [reply test/utils/typography_utils_test.dart:76]: Thats a good question

**Review by millerovv — APPROVED** [review]

**Pull request merged**